core/src/modules/caches/sourcesCache.ts

```diff
@@ -0,0 +1,64 @@
+import { ModelCache } from "../modelCache";
+import { Source } from "../../models/Source";
+import { Option } from "../../models/Option";
+import { Mapping } from "../../models/Mapping";
+import { Includeable, WhereAttributeHash } from "sequelize";
+
+function getInclude() {
+  const include: Includeable[] = [
+    { model: Option, required: false },
+    { model: Mapping, required: false },
+  ];
+  return include;
+}
+
+async function findAllWithCache(
+  this: ModelCache<Source>,
+  modelId?: string,
+  state?: Source["state"]
+) {
+  const now = new Date().getTime();
+  if (this.expires > now && this.instances.length > 0) {
+    return modelId
+      ? this.instances.filter(
+          (s) => s?.modelId === modelId && (!state || s.state === state)
+        )
+      : this.instances.filter((s) => !state || s.state === state);
+  } else {
+    this.instances = await Source.unscoped().findAll({ include: getInclude() });
+    this.expires = now + this.TTL;
+    return modelId
+      ? this.instances.filter(
+          (s) => s?.modelId === modelId && (!state || s.state === state)
+        )
+      : this.instances.filter((s) => !state || s.state === state);
+  }
+}
+
+async function findOneWithCache(
+  this: ModelCache<Source>,
+  value: string,
+  modelId?: string,
+  state?: Source["state"],
+  lookupKey: keyof Source = "id"
+) {
+  const instances = await this.findAllWithCache(modelId);
+  let instance = instances.find((i) => i[lookupKey] === value);
+
+  if (!instance) {
+    const where: WhereAttributeHash = { [lookupKey]: value };
+    if (state) where.state = state;
+    instance = await Source.unscoped().findOne({
+      where,
+      include: getInclude(),
+    });
+    if (instance) this.invalidate();
+  }
+
+  return instance;
+}
```

Comment:
I'm seeing this repeated a lot. I think we might be able to improve this pattern.
An idea:
```
const generateFindOneWithCache = <T>(include: Includable[]) => {
return async function findOneWithCache(
  this: ModelCache<T>,
  value: string,
  modelId?: string,
  state?: T["state"],
  lookupKey: keyof T = "id"
) {
  const instances = await this.findAllWithCache(modelId);
  let instance = instances.find((i) => i[lookupKey] === value);

  if (!instance) {
    const where: WhereAttributeHash = { [lookupKey]: value };
    if (state) where.state = state;
    instance = await Property.unscoped().findOne({
      where,
      include,
    });
    if (instance) this.invalidate();
  }

  return instance;
}
}
```

I haven't pulled it down and tried this but I think it's worth a try.

Reply:
Maybe I need to pull this down and dig into it since I see a class generic below.